presto-function-namespace-managers/src/main/java/com/facebook/presto/sqlfunction/db/FunctionNamespaceDao.java

```diff
@@ -0,0 +1,176 @@
+import com.facebook.presto.spi.type.TypeSignature;
+import org.jdbi.v3.sqlobject.config.RegisterArgumentFactories;
+import org.jdbi.v3.sqlobject.config.RegisterArgumentFactory;
+import org.jdbi.v3.sqlobject.config.RegisterRowMapper;
+import org.jdbi.v3.sqlobject.customizer.Bind;
+import org.jdbi.v3.sqlobject.customizer.Define;
+import org.jdbi.v3.sqlobject.statement.SqlQuery;
+import org.jdbi.v3.sqlobject.statement.SqlUpdate;
+
+import java.util.List;
+import java.util.Optional;
+
+@RegisterRowMapper(SqlInvokedFunctionRowMapper.class)
+@RegisterArgumentFactories({
+        @RegisterArgumentFactory(QualifiedFunctionNameArgumentFactory.class),
+        @RegisterArgumentFactory(SqlFunctionIdArgumentFactory.class),
+        @RegisterArgumentFactory(SqlParametersArgumentFactory.class),
+        @RegisterArgumentFactory(TypeSignatureArgumentFactory.class),
+})
+public interface FunctionNamespaceDao
+{
+    @SqlUpdate("CREATE TABLE `<table_name>` (\n" +
+            "  `id` bigint(20) NOT NULL AUTO_INCREMENT,\n" +
+            "  `function_id` varchar(700) NOT NULL,\n" +
+            "  `version` bigint(20) unsigned NOT NULL,\n" +
+            "  `function_name` varchar(512) NOT NULL,\n" +
+            "  `parameters` varchar(2048) NOT NULL,\n" +
+            "  `return_type` varchar(128) NOT NULL,\n" +
+            "  `language` varchar(64) NOT NULL,\n" +
+            "  `deterministic` tinyint(1) NOT NULL,\n" +
+            "  `called_on_null_input` tinyint(1) NOT NULL,\n" +
+            "  `body` mediumtext,\n" +
+            "  `description` text,\n" +
+            "  `created_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,\n" +
+            "  `modified_time` TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,\n" +
+            "  PRIMARY KEY (`id`),\n" +
+            "  UNIQUE KEY `function_id_version` (`function_id`,`version`),\n" +
+            "  KEY `function_name` (`function_name`)\n" +
+            ")")
+    void createSqlInvokedFunctionsTable(@Define("table_name") String tableName);
+
+    @SqlQuery("SELECT\n" +
+            "    t.`function_name`,\n" +
+            "    t.`parameters`,\n" +
+            "    t.`return_type`,\n" +
+            "    t.`description`,\n" +
+            "    t.`language`,\n" +
+            "    t.`deterministic`,\n" +
+            "    t.`called_on_null_input`,\n" +
+            "    t.`body`,\n" +
+            "    t.`version`\n" +
+            "FROM `<table_name>` t\n" +
+            "JOIN (\n" +
+            "    SELECT\n" +
+            "        `function_id`,\n" +
+            "        MAX(`version`) `version`\n" +
+            "    FROM `<table_name>`\n" +
+            "    GROUP BY\n" +
+            "        `function_id`\n" +
+            ") v\n" +
+            "    ON t.`function_id` = v.`function_id`\n " +
+            "    AND t.`version` = v.`version`")
+    List<SqlInvokedFunction> listFunctions(@Define("table_name") String tableName);
```

Comment:
I think `listFunctions` only need to return a `SqlFunction`. This is used for `SHOW FUNCTIONS`, which only needs the metadata, not the body, or the version. If you cannot find a good class, maybe just return `List<FunctionMetadata>`.

Reply:
This `listFunctions` is at the end of the call stack for `Metadata.listFunction()`. We'll need refactor
a few layers of methods in order to do that, including `Metadata.listFunction()`, `FunctionManager.listFunction()`, `FunctionNamespaceManager.listFunction()`. For now, I'm adding a TODO in `FunctionNamespaceManager.listFunction()`